src/cache/local.ts

```diff
@@ -14,47 +14,145 @@ export interface LocalCacheEntry<T> {
+  prev: LinkedListNode | null
+  key: string
+  constructor(key: string, data: T) {
+    this.data = data
+    this.next = null
+    this.prev = null
+    this.key = key
+  }
+}
+
+class DoubleLinkedList {
+  head: LinkedListNode | null
+  tail: LinkedListNode | null
+  size: number
+  constructor() {
+    this.head = null
+    this.tail = null
+    this.size = 0
+  }
+
+  insertAtTail(node: LinkedListNode) {
+    if (!this.tail) {
+      this.tail = node
+      this.head = node
+      node.next = null
+    } else {
+      this.tail.next = node
+      node.prev = this.tail
+      this.tail = node
+      node.next = null
+    }
+
+    this.size++
+    return node
+  }
+
+  remove(node: LinkedListNode | null): LinkedListNode | undefined {
+    if (!node) {
+      return
+    }
+
+    if (node.prev !== null) {
+      node.prev.next = node.next
+    }
+
+    if (node.next !== null) {
+      node.next.prev = node.prev
+    }
+
+    if (node === this.head) {
+      this.head = this.head.next
+    }
+
+    if (node === this.tail) {
+      this.tail = this.tail.prev
+    }
+    this.size--
+    return node
+  }
+
+  removeHead(): LinkedListNode | undefined {
+    return this.remove(this.head)
+  }
+}
+
 /**
- * Local implementation of a Cache. It uses a simple js Object, storing entries with both
- * a value and an expiration timestamp. Expired entries are deleted on reads (i.e. no background gc/upkeep).
- *
+ * Local implementation of a Cache. It uses a DoubleLinkedList class for storing cache nodes.
+ * Each LinkedListNode is storing entries with both value and an expiration timestamp.
+ * Expired entries are deleted on reads (i.e. no background gc/upkeep).
  * @typeParam T - the type for the entries' values
  */
 export class LocalCache<T = unknown> implements Cache<T> {
   type = CacheTypes.Local
+  capacity: number
+  // Cache will hold cache keys as 'key' and references to LinkedListNodes as 'value' for fast search
+  cache: Map<string, LinkedListNode<LocalCacheEntry<T>>>
+  list: DoubleLinkedList
 
-  store: Record<string, LocalCacheEntry<T>> = {}
+  constructor(capacity: number) {
+    this.capacity = capacity
+    this.cache = new Map()
+    this.list = new DoubleLinkedList()
+  }
 
   async get(key: string): Promise<T | undefined> {
     logger.trace(`Getting key ${key}`)
-    const entry = this.store[key]
-
-    if (!entry) {
+    if (this.cache.has(key)) {
+      const node = this.cache.get(key) as LinkedListNode<LocalCacheEntry<T>>
+      const expired = node.data.expirationTimestamp <= Date.now()
+      if (expired) {
+        logger.debug('Entry in local cache expired, deleting and returning undefined')
+        this.delete(key)
+        return undefined
+      } else {
+        logger.debug('Found valid entry in local cache, returning value')
+        // When key is accessed we move it to end of the list as we keep recently used entries there
+        this.moveToTail(node)
+        return node.data.value
+      }
```

Comment:
@alejoberardino  Is it ok that in this case that some frequently read cache entries might be removed from cache before they are expired ?

Reply:
Yes, it's ok if that happens. But what we should do is add a `warn` log when an entry is evicted because of a size limitation; this limit is a precaution to avoid DoS but it should never be hit in practice